src/utils/memoryCard.lua

```diff
@@ -0,0 +1,160 @@
+local pd <const> = playdate
+local ds <const> = pd.datastore
+
+class("MemoryCard").extends()
+
+local FILE_NAME <const> = "discontrolled_save_data"
+
+local DEFAULT_DATA <const> = {
+  lastPlayed = { world = 1, level = 1 },
+  worlds = {
+    [1] = {
+      levels = {
+        [1] = {
+          total = 0,
+          rescued = 0
+        },
+        [2] = {
+          total = 0,
+          rescued = 0
+        },
+        [3] = {
+          total = 0,
+          rescued = 0
+        },
+        [4] = {
+          total = 0,
+          rescued = 0
+        },
+      }
+    },
+    [2] = {
+      levels = {
+        [1] = {
+          total = 0,
+          rescued = 0
+        },
+        [2] = {
+          total = 0,
+          rescued = 0
+        },
+        [3] = {
+          total = 0,
+          rescued = 0
+        },
+        [4] = {
+          total = 0,
+          rescued = 0
+        },
+      }
+    },
+    [3] = {
+      levels = {
+        [1] = {
+          total = 0,
+          rescued = 0
+        },
+        [2] = {
+          total = 0,
+          rescued = 0
+        },
+        [3] = {
+          total = 0,
+          rescued = 0
+        },
+        [4] = {
+          total = 0,
+          rescued = 0
+        },
+      }
+    },
+    [4] = {
+      levels = {
+        [1] = {
+          total = 0,
+          rescued = 0
+        },
+        [2] = {
+          total = 0,
+          rescued = 0
+        },
+        [3] = {
+          total = 0,
+          rescued = 0
+        },
+        [4] = {
+          total = 0,
+          rescued = 0
+        },
+      }
+    },
+  }
+}
+
+-- local functions, actual data access
+
+local function saveData(data)
+  assert(data, "tried to save nil data")
+  ds.write(data, FILE_NAME);
+end
+
+local function loadData()
+  local data = ds.read(FILE_NAME)
+  if (data) then return data end
+  return DEFAULT_DATA
+end
+
+-- local functions, helpers
+
+local function assertValidInput(world, level)
+  assert(world > 0 and world < 5, "world must be between 1-4 (inclusive)")
+  assert(level > 0 and level < 5, "level must be between 1-4 (inclusive)")
+end
+
+-- static functions, to be called by other classes
+
+-- sets the total bots for the level
+-- ideally called during LDtk loading so
+-- it can be done dynamically
+function MemoryCard.setTotalBotsForLevel(world, level, total)
```

Comment:
If we're going to hard-code the level data, we shouldn't need this. Not sure setting this dynamically is an efficient solution